weblate/integrations/models.py

```diff
@@ -0,0 +1,143 @@
+# Copyright © Michal Čihař <michal@weblate.org>
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from django.db import models
+from django.db.models import Q
+from django.urls import reverse
+from django.utils.functional import cached_property
+
+from weblate.addons.events import EVENT_CHOICES
+from weblate.trans.models import Change, Component
+from weblate.utils.classloader import ClassLoader
+from weblate.utils.fields import JSONField
+
+# Initialize integrations registry
+INTEGRATIONS = ClassLoader("WEBLATE_INTEGRATIONS", False)
+
+
+class IntegrationQuerySet(models.QuerySet):
+    def filter_component(self, component):
+        return self.prefetch_related("event_set").filter(
+            (Q(component=component) & Q(project_scope=False))
+            | (Q(component__project=component.project) & Q(project_scope=True))
+            | (Q(component__linked_component=component) & Q(repo_scope=True))
+            | (Q(component=component.linked_component) & Q(repo_scope=True))
+        )
+
+    def filter_event(self, component, event):
+        # To do: Add integrations cache property in component
+        return component.inegrations_cache[event]
+
+
+class Integration(models.Model):
+    component = models.ForeignKey(Component, on_delete=models.deletion.CASCADE)
+    name = models.CharField(max_length=100)
+    configuration = JSONField()
+    state = JSONField()
+    project_scope = models.BooleanField(default=False, db_index=True)
+    repo_scope = models.BooleanField(default=False, db_index=True)
```

Comment:
```suggestion
```

These are probably not a good idea here – in the add-ons, the scoping is defined by the add-on, here it will be user defined. Instead, add a foreign key to `Project` as well, and make both foreign keys nullable.